### ISSUE TITLE
An external user can modify admin permissions

### ISSUE THREAD

**Issue by c3do** [Dolibarr/dolibarr#27976]
### Bug

An external user can receive all permissions related to users and user groups. He can therefore update his own authorizations, but also the administrator's permissions.

### Environment Version

18

### Expected and actual behavior

An external user should not be able to receive certain permissions

### Steps to reproduce the behavior

1. Create an external user
2. Give him all permissions related to users
3. Log in with external user
4. Navigate to Home > Users & Groups > List of users > The external user's card
5. Navigate to the previous cards by clicking on the “<” arrow to the admin card
6. Change permissions

**Comment by altairis-tof** [comment]
"Give him all permissions related to users" ???

**Comment by c3do** [comment]
@altairis-tof 
![Screenshot 2024-02-08 09 59 56](https://github.com/Dolibarr/dolibarr/assets/35066297/7d94e9c0-d1fd-4f1f-8b1b-5073f3907f10)

**Comment by altairis-tof** [comment]
but why do you give him these permissions ?

**Comment by c3do** [comment]
> but why do you give him these permissions ?

but why would it be permissible to give these permissions?

**Comment by altairis-tof** [comment]
good point, but this is not a bug. 
when you define external user you have to define permissions accordingly to what you want.
maybe there is a usecase when these permissions have to be setable for external user ?

**Comment by altairis-tof** [comment]
after more thinking i agree with you; do you think you can do the fix PR ? @c3do

**Comment by altairis-tof** [comment]
i think the fix would be to forbid to define sensible permissions for external users.
but i'd be glad to have @eldy point of view before we code anything.

**Comment by eldy** [comment]
By default an external user is always limited to its company (this is definition of external user so this rule should not be bypassed).
So if we give him permission to edit users, he must have permission to edit users of only its company too (so other external users of the same company).

**Comment by daraelmin** [comment]
if permissions are limited in the future for external users, it will absolutely be necessary to correct the bug on the creation of an internal user from a member (see below). 

For the moment, this poses little problem since we can grant them all the permissions, but if in the future we can no longer...

https://github.com/Dolibarr/dolibarr/issues/26392

**Comment by JonBendtsen** [comment]
> if permissions are limited in the future for external users, it will absolutely be necessary to correct the bug on the creation of an internal user from a member (see below).
> 
> For the moment, this poses little problem since we can grant them all the permissions, but if in the future we can no longer...
> 
> #26392

really? all permissions are automatically granted?

**Comment by daraelmin** [comment]
Permission are not automaticaly granted BUT it is possible to give them to a member who becomes a user, despite the fact that you can't turn a member into an internal user.

If it become impossible to give permission to an external user, we must offer the possibility to turn a member into an intenal user.

**Comment by JonBendtsen** [comment]
> Permission are not automaticaly granted BUT it is possible to give them to a member who becomes a user, despite the fact that you can't turn a member into an internal user.
> 
> If it become impossible to give permission to an external user, we must offer the possibility to turn a member into an intenal user.

okay, that makes more sense. How come Dolibarr distinguishes between internal and external users?

**Comment by daraelmin** [comment]
If a user is linked to a thirdparty, it become an external user.

It is therefore possible to create an internal user from a member as long as they are not linked to a third party... but if they have to pay a membership fee, especially if online payment is possible or if accounting is enabled, there is a good chance that members will be linked to a third party.

For your information, once a member who is not linked to a third party becomes an internal user, it is possible to link the member to a third party without worry.

I'm not sure where this comes from.

**Comment by JonBendtsen** [comment]
> If a user is linked to a thirdparty, it become an external user.
> 
> It is therefore possible to create an internal user from a member as long as they are not linked to a third party... but if they have to pay a membership fee, especially if online payment is possible or if accounting is enabled, there is a good chance that members will be linked to a third party.
> 
> For your information, once a member who is not linked to a third party becomes an internal user, it is possible to link the member to a third party without worry.

All our members have a linked thirdparty, but not all thirdparties are members.

Some of our members have a login so they can help run our organisation, like volunteers who might have an expense report and also our teachers so they can accept people into our classes.

We are a non profit dance organisation, we use dolibarr and a python flask form for registering for dance classes. Then from inside Dolibarr we validate orders and they are sent an email to pay, and then dolibarr auto creates an invoice.

**Comment by daraelmin** [comment]
I must correct my self:

An external user is a user directly linked to a thirdparty. So if a user is linked to a member wich is linked to a thirdparty, the user is internal.

Actualy, it is impossible to remoye the link between user and thirdparty without using SQL and it is impossible zo create a user from a member linked to a thirdparty without creating a link between the new user and the thirdparty member.

**Comment by bos4711** [comment]
`An external user is a user directly linked to a thirdparty. `

What is the point of external users?

Or let me rephrase that: Let's say person X is listed as contact for third party Y. What purpose does it serve for anyone or anything that person X being an external user logs in to our Dolibarr?

**Comment by daraelmin** [comment]
First point : some thirdparty are not company. For example, a member is a thirdparty "physical person" which has no contact.

Second point : most Dolibarr administrators need their customers, external users, to be able to access their data, their customer account, to download invoices, check orders, track orders, etc.

**Comment by JonBendtsen** [comment]
> First point : some thirdparty are not company. For example, a member is a thirdparty "physical person" which has no contact.
> 
> Second point : most Dolibarr administrators need their customers, external users, to be able to access their data, their customer account, to download invoices, check orders, track orders, etc.

yes all my thirdparties are private persons.

That access could be given without using a username/password, this could be a unique link. But even if they do have a user, then it could be some permissions that was given to a user if they have access to their own thirdparty data.

**Comment by daraelmin** [comment]
Some of this data may be confidential and/or regulated (rgpd et consort), such as postal addresses, emails, first and last names, telephones, etc.

It doesn't seem right to display them via a public link. I'm already rather uncomfortable with the public page for online renewal and payment of members' subscriptions (in the case of a trade union, we need to be able to guarantee people's anonymity, the same goes for members of political parties).

If, in addition, customers are allowed to update their data....

No, a password-protected personal space seems essential to me.

**Comment by JonBendtsen** [comment]
> Some of this data may be confidential and/or regulated (rgpd et consort), such as postal addresses, emails, first and last names, telephones, etc.
> 
> It doesn't seem right to display them via a public link. I'm already rather uncomfortable with the public page for online renewal and payment of members' subscriptions (in the case of a trade union, we need to be able to guarantee people's anonymity, the same goes for members of political parties).
> 
> If, in addition, customers are allowed to update their data....
> 
> No, a password-protected personal space seems essential to me.

I don't mean a public link, and Dolibarr already shares the order, proposal and invoices using a unique link, so it can't be a totally wrong solution.

**Comment by eldy** [comment]
To summarize all the thread:

By default an external user is always limited to its company (this is definition of external user so this rule should not be bypassed).
So if we give him permission to edit users, he must have permission to edit users of only its company too (so other external users of the same company). If external user can do more, it is a bug and must be fixed to be limited to external users in same company only.
